phase4-entsog-client/src/main/java/com/helger/phase4/entsog/Phase4ENTSOGSender.java

```diff
@@ -0,0 +1,191 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.helger.phase4.entsog;
+
+import java.nio.charset.Charset;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.OverridingMethodsMustInvokeSuper;
+import javax.annotation.concurrent.Immutable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.helger.phase4.attachment.Phase4OutgoingAttachment;
+import com.helger.phase4.attachment.WSS4JAttachment;
+import com.helger.phase4.client.AS4ClientUserMessage;
+import com.helger.phase4.crypto.ECryptoKeyIdentifierType;
+import com.helger.phase4.sender.AS4BidirectionalClientHelper;
+import com.helger.phase4.sender.AbstractAS4UserMessageBuilder;
+import com.helger.phase4.util.AS4ResourceHelper;
+import com.helger.phase4.util.Phase4Exception;
+
+/**
+ * This class contains all the specifics to send AS4 messages with the ENTSOG
+ * profile. See <code>sendAS4Message</code> as the main method to trigger the
+ * sending, with all potential customization.
+ *
+ * @author Pavel Rotek
+ * @since 0.13.3
+ */
+@Immutable
+public final class Phase4ENTSOGSender {
+	private static final Logger LOGGER = LoggerFactory.getLogger(Phase4ENTSOGSender.class);
+
+	private Phase4ENTSOGSender() {
+	}
+
+	/**
+	 * @return Create a new Builder for AS4 messages if the payload is present.
+	 *         Never <code>null</code>.
+	 */
+	@Nonnull
+	public static ENTSOGUserMessageBuilder builder() {
+		return new ENTSOGUserMessageBuilder();
+	}
+
+	/**
+	 * Abstract ENTSOG UserMessage builder class with sanity methods
+	 *
+	 * @author Philip Helger
+	 * @param <IMPLTYPE> The implementation type
+	 */
+	public abstract static class AbstractENTSOGUserMessageBuilder<IMPLTYPE extends AbstractENTSOGUserMessageBuilder<IMPLTYPE>>
+			extends AbstractAS4UserMessageBuilder<IMPLTYPE> {
+
+		private ECryptoKeyIdentifierType m_signingKeyIdentifierType = ECryptoKeyIdentifierType.ISSUER_SERIAL;
+		private ECryptoKeyIdentifierType m_encryptionKeyIdentifierType = ECryptoKeyIdentifierType.ISSUER_SERIAL;
+		
+		private Phase4OutgoingAttachment m_aPayload;
+		private Charset m_aPayloadCharset;
+		private String m_aPayloadDocumentType;
+		
+		protected AbstractENTSOGUserMessageBuilder() {
+			// Override default values
+			try {
+				httpClientFactory(new Phase4ENTSOGHttpClientSettings());
+			} catch (final Exception ex) {
+				throw new IllegalStateException("Failed to init AS4 Client builder", ex);
+			}
+		}
+
+		/**
+		 * Encryption Key Identifier Type.
+		 *
+		 * @param keyIdentifier ECryptoKeyIdentifierType. Defaults to ISSUER_SERIAL
+		 * @return this for chaining
+		 */
+		public final IMPLTYPE setEncryptionKeyIdentifierType(ECryptoKeyIdentifierType encryptionKeyIdentifierType) {
```

Comment:
Hm - does it make sense to instead provide a setter for the `AS4CryptParams`?

Reply:
I gave it a second thought and I think it should stay as it is. Because if we take it up the hierarchy, this leads to the problem that users of other profiles might accidentally use it. If this is the only parameter that should be changeable - so be it.

Reply:
I feel setting the encryption/signature type (ISSUER_TOKEN vs. BINARY_SECURITY_TOKEN) is handful for all profiles. maybe move this method up to abstract class?!?! ENTSOG/AS4 doesn't specify whether to use ISSUER_TOKEN or BST or other WS-Security way.  I think it's left to the business partners decission which one they will use.  I can imagine that also other parts of cryptParams() can be set.

Reply:
The implication would be, that "Peppol" users could change the KeyIdentifierType even though they are not supposed to do so ;-) - but than that needs to be verified by the validator hmmm.... I'd prefer to leave it here to avoid avoidable user errors ;-)